ktor-http/common/src/io/ktor/http/URLParser.kt

```diff
@@ -89,34 +99,72 @@ internal fun URLBuilder.takeFromUnsafe(urlString: String): URLBuilder {
         // overwrite the path
         ""
     }
+
     val pathEnd = urlString.indexOfAny("?#".toCharArray(), startIndex).takeIf { it > 0 } ?: endIndex
     val rawPath = urlString.substring(startIndex, pathEnd)
+
     encodedPath += rawPath.encodeURLPath()
     startIndex = pathEnd
 
     // Query
     if (startIndex < endIndex && urlString[startIndex] == '?') {
-        if (startIndex + 1 == endIndex) {
-            trailingQuery = true
-            return this
-        }
+        startIndex = parseQuery(urlString, startIndex, endIndex)
+    }
+
+    // Fragment
+    parseFragment(urlString, startIndex, endIndex)
+    return this
+}
 
-        val fragmentStart = urlString.indexOf('#', startIndex + 1).takeIf { it > 0 } ?: endIndex
+private fun URLBuilder.parseFile(urlString: String, startIndex: Int, endIndex: Int, slashCount: Int) {
+    when (slashCount) {
+        2 -> {
+            val nextSlash = urlString.indexOf('/', startIndex)
+            if (nextSlash == -1 || nextSlash == endIndex) {
+                host = urlString.substring(startIndex, endIndex)
+                return
+            }
 
-        val rawParameters = parseQueryString(urlString.substring(startIndex + 1, fragmentStart))
-        rawParameters.forEach { key, values ->
-            parameters.appendAll(key, values)
+            host = urlString.substring(startIndex, nextSlash)
+            encodedPath = urlString.substring(nextSlash, endIndex)
         }
+        3 -> {
+            host = ""
+            encodedPath = "/" + urlString.substring(startIndex, endIndex)
+        }
+    }
+}
 
-        startIndex = fragmentStart
+private fun URLBuilder.parseMailto(urlString: String, startIndex: Int, endIndex: Int) {
+    val delimiter = urlString.indexOf("@", startIndex);
+    if (delimiter == -1) {
+        throw IllegalArgumentException("Invalid mailto url: $urlString, it should contains '@'.")
     }
 
-    // Fragment
+    user = urlString.substring(startIndex, delimiter)
```

Comment:
Shouldn't we decode/de-escape user and password here?

Reply:
As I get it, we store everything in escaped format(the same as for regular URLs)

Reply:
Then we shouldn't escape them when formatting otherwise it breaks parse-then-encode.

Reply:
fixed as discussed, please take a look